certora/specs/RecoveryConfirmationSignatureValidity.spec

```diff
@@ -0,0 +1,131 @@
+using SafeHarness as safeContract;
+
+methods {
+    // Social Recovery Module Functions
+    function nonce(address) external returns (uint256) envfree;
+    function isGuardian(address, address) external returns (bool) envfree;
+
+    // Social Recovery Module Summaries
+    function getRecoveryHash(address, address[] calldata, uint256, uint256) internal returns (bytes32) => CONSTANT;
+    function SignatureChecker.isValidSignatureNow(address signer, bytes32 dataHash, bytes memory signatures) internal returns (bool) => isValidSignatureNowSummary(signer, dataHash, signatures);
+
+    // Wildcard Functions
+    function _.execTransactionFromModule(address to, uint256 value, bytes data, Enum.Operation operation) external with (env e) => summarizeSafeExecTransactionFromModule(calledContract, e, to, value, data, operation) expect bool ALL;
+    function _.isModuleEnabled(address module) external => DISPATCHER(false);
+    function _.isOwner(address owner) external => DISPATCHER(false);
+    function _.getOwners() external => DISPATCHER(false);
+    function _._ external => DISPATCH[] default NONDET;
+}
+
+
+// A summary function that helps the prover resolve calls to `safeContract`.
+function summarizeSafeExecTransactionFromModule(address callee, env e, address to, uint256 value, bytes data, Enum.Operation operation) returns bool {
+    if (callee == safeContract) {
+        return safeContract.execTransactionFromModule(e, to, value, data, operation);
+    }
+    return _;
+}
+
+ghost isValidSignatureNowSummary(address, bytes32, bytes) returns bool;
+
+persistent ghost mathint recoveryConfirmationCount {
+    init_state axiom recoveryConfirmationCount == 0;
+}
+
+hook Sstore confirmedHashes[KEY bytes32 recoveryHash][KEY address guardian] bool value {
+    recoveryConfirmationCount = recoveryConfirmationCount + 1;
+}
+
+// This rule the `multiConfirmRecovery` function can only be called with legitimate signatures.
+// It assumes that the harnessed `checkSignatures` function is implemented correctly and reverts if the signatures are invalid.
+rule multiConfirmRecoveryOnlyWithLegitimateSignatures(env e) {
+    address _wallet;
+    address[] _newOwners;
+    uint256 _newThreshold;
+    uint256 walletNonce = currentContract.nonce(_wallet);
+    SocialRecoveryModule.SignatureData[] signatures;
+    bool _execute;
+    bytes32 recoveryHash = getRecoveryHash(
+        e,
+        _wallet,
+        _newOwners,
+        _newThreshold,
+        walletNonce
+    );
+
+    checkSignatures@withrevert(e, _wallet, recoveryHash, signatures);
+    bool signatureCheckSuccess = !lastReverted;
+
+    multiConfirmRecovery(e, _wallet, _newOwners, _newThreshold, signatures, _execute);
+
+    assert signatureCheckSuccess, "Recovery confirmed with invalid signatures";
```

Comment:
Don't know if its worth mentioning in a comment, but this implies that there are cases where `multiConfirmRecovery` reverts, but `checkSignatures` does not. If there are no such cases, then you should assert equivalence between `signatureCheckSuccess` and `multiConfirmRecoverySuccess`.

Reply:
I considered this when writing the rule and decided it was too big of an implementation detail and might be excessive. The main thing I wanted to verify is that if the signature fails, the function always reverts.

I tried asserting the equivalence and it works: https://prover.certora.com/output/6575/f1dbf26deea845f9acfa5a5ce96fb59e?anonymousKey=4e31370b9350f69841985a658b169be5f80d2e39

Let me know if I should change it